[FLINK-16200][sql] Support JSON_EXISTS for blink planner

XuQianJin-Stars opened the pull request:
## What is the purpose of the change

Support `JSON_EXISTS` function for blink planner

## Brief change log

*(for example:)*

- Introduce `JSON_EXISTS` to `FlinkSqlOperatorTable`
- Add corresponding test cases

## Verifying this change

*(Please pick either of the following options)*

This change added tests in `ScalarFunctionsTest.scala` 

*(example:)*

- *Added integration tests for end-to-end deployment with large payloads (100MB)*
- *Extended integration test for recovery after master (JobManager) failure*
- *Added test that validates that TaskInfo is transferred only once across recoveries*
- *Manually verified the change by running a 4 node cluser with 2 JobManagers and 4 TaskManagers, a stateful streaming program, and killing one JobManager and two TaskManagers during the execution, verifying that recovery happens correctly.*

## Does this pull request potentially affect one of the following parts:

- Dependencies (does it add or upgrade a dependency): (yes / no)
- The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no)
- The serializers: (yes / no / don't know)
- The runtime per-record code paths (performance sensitive): (yes / no / don't know)
- Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / no / don't know)
- The S3 file system connector: (yes / no / don't know)

## Documentation

- Does this pull request introduce a new feature? (yes / no)
- If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

libenchao commented on flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/calls/BuiltInMethods.scala:
```diff
@@ -22,11 +22,13 @@ import org.apache.flink.table.dataformat.{Decimal, SqlTimestamp}
 import java.lang.{Byte => JByte, Integer => JInteger, Long => JLong, Short => JShort}
 import java.util.TimeZone
 
+import org.apache.calcite.sql.SqlJsonExistsErrorBehavior
```
import needs to be ordered alphabetically.

libenchao commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java:
```diff
@@ -674,6 +674,14 @@
+		new BuiltInFunctionDefinition.Builder()
+			.name("json_exists")
+			.kind(SCALAR)
+			.outputTypeStrategy(TypeStrategies.MISSING)
```
It's weird, why not directly specify the outputType? I know that all function definitions are all defined with `MISSING` strategy, and it make me more confuse.

XuQianJin-Stars replied:
> It's weird, why not directly specify the outputType? I know that all function definitions are all defined with `MISSING` strategy, and it make me more confuse.

The type specified here becomes a cast.

libenchao commented on flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/calls/FunctionGenerator.scala:
```diff
@@ -22,10 +22,8 @@ import org.apache.flink.table.planner.functions.sql.FlinkSqlOperatorTable._
-
 import org.apache.calcite.sql.SqlOperator
 import org.apache.calcite.util.BuiltInMethod
-
```
don't need to remove these blank lines.

libenchao commented on flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/ScalarFunctionsTest.scala:
```diff
@@ -4195,4 +4195,48 @@ class ScalarFunctionsTest extends ScalarTypesTestBase {
+  // JSON functions
+  //-------------------------------------------------------------------
+  @Test
+  def testJsonExists(): Unit = {
```
These tests actually test only SQL expressions, not table api expressions. Have you ever tried to run those expressions from Table API?

XuQianJin-Stars replied:
> These tests actually test only SQL expressions, not table api expressions. Have you ever tried to run those expressions from Table API?

@wuchong  recommends just SQL  expressions first.

libenchao replied:
Do you mean supporting these json functions in SQL first? Or just testing for SQL?
If the answer is supporting these json functions in SQL firstly, I prefer we do not add these table api related expressions for now, which are not tested for now.

XuQianJin-Stars replied:
> I prefer we do not add these table api related expressions for now,

You are right, Let me do not add these table api related expressions for now.

XuQianJin-Stars commented:
hi, @libenchao I addressed it.

libenchao commented:
One more tip, you don't need to force push your branch each time when addressing comments, it will be hard for the reviewers to track changes.

XuQianJin-Stars commented:
> One more tip, you don't need to force push your branch each time when addressing comments, it will be hard for the reviewers to track changes.

ok

libenchao commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java:
```diff
@@ -674,6 +674,14 @@
 			.build();
 
+	// json functions
+	public static final BuiltInFunctionDefinition JSON_EXISTS =
```
we don't need to add it here if we only aim to SQL.

XuQianJin-Stars replied:
> we don't need to add it here if we only aim to SQL.

well, I removed it.

libenchao commented on flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/calls/FunctionGenerator.scala:
```diff
@@ -801,6 +801,20 @@ object FunctionGenerator {
 
+  addSqlFunctionMethod(
+    JSON_EXISTS,
+    Seq(CHAR, CHAR),
```
What about `Seq(VARCHAR, CHAR)`/`Seq(CHAR, VARCHAR)`?

XuQianJin-Stars replied:
> What about `Seq(VARCHAR, CHAR)`/`Seq(CHAR, VARCHAR)`?

```
  public SqlJsonExistsFunction() {
    super("JSON_EXISTS", SqlKind.OTHER_FUNCTION,
        ReturnTypes.cascade(ReturnTypes.BOOLEAN, SqlTypeTransforms.FORCE_NULLABLE), null,
        OperandTypes.or(
            OperandTypes.family(SqlTypeFamily.ANY, SqlTypeFamily.CHARACTER),
            OperandTypes.family(SqlTypeFamily.ANY, SqlTypeFamily.CHARACTER, SqlTypeFamily.ANY)),
        SqlFunctionCategory.SYSTEM);
  }
```
The `path` parameter can only be CHAR.  This need to add `Seq(VARCHAR, CHAR)`

libenchao replied:
According to SQL 2016 Standard: 
```
<JSON API common syntax> ::= <JSON context item> <comma> <JSON path specification> [ AS <JSON table path name> ] [ <JSON passing clause> ]
<JSON context item> ::= <JSON value expression>
<JSON path specification> ::= <character string literal>
```
Second parameter can only be string literal, and in Flink, it's `CHAR` type.

XuQianJin-Stars replied:
> According to SQL 2016 Standard:
> 
> ```
> <JSON API common syntax> ::= <JSON context item> <comma> <JSON path specification> [ AS <JSON table path name> ] [ <JSON passing clause> ]
> <JSON context item> ::= <JSON value expression>
> <JSON path specification> ::= <character string literal>
> ```
> 
> Second parameter can only be string literal, and in Flink, it's `CHAR` type.

Thank you very much for your analysis, it has helped me a lot.

libenchao commented:
LGTM, +1 for merging.

wuchong commented on flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/JsonFunctionsTest.scala:
```diff
@@ -125,4 +127,53 @@ class JsonFunctionsTest extends ExpressionTestBase {
   }
 
+  @Test
+  def testJsonExists(): Unit = {
```
Could you add a more complex test? Checking existence on a nested fields? e.g. `json_exist(f0, 'lax $.info.address[0].state')`. You can take [Sql Server](https://docs.microsoft.com/en-us/sql/t-sql/functions/json-value-transact-sql?view=sql-server-ver15#lax-mode-and-strict-mode) as an example.

wuchong commented on flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/JsonFunctionsTest.scala:
```diff
@@ -125,4 +127,53 @@ class JsonFunctionsTest extends ExpressionTestBase {
+  @Test
+  def testJsonFuncError(): Unit = {
+    expectedException.expect(classOf[CodeGenException])
+    expectedException.expectMessage(startsWith("Unsupported call: JSON_EXISTS"))
```
This exception message is still misleading. We already support `JSON_EXISTS`, why the exception says not? I think we should improve the exception to give a better understandable message, e.g. `the json path 'lax $' is illegal.`

XuQianJin-Stars replied:
> This exception message is still misleading. We already support `JSON_EXISTS`, why the exception says not? I think we should improve the exception to give a better understandable message, e.g. `the json path 'lax $' is illegal.`

Because `JSON_EXISTS (INT, CHAR (5) NOT NULL, RAW ('org.apache.calcite.sql.SqlJsonExistsErrorBehavior',?)` This is not supported. The json path 'lax $' is illegal will return `false`.

wuchong replied:
I got what you want to throw for the exception. Currently, Calcite JSON_EXISTS definition accepts any type for first argument, however, we only support string in the code generation (do you know why Calcite accept any type?)

For this purpose, I would suggest to define JSON_EXISTS ourselves:

```
	public static final SqlFunction JSON_EXISTS = new SqlFunction(
		"JSON_EXISTS",
		SqlKind.OTHER_FUNCTION,
		ReturnTypes.cascade(ReturnTypes.BOOLEAN, SqlTypeTransforms.FORCE_NULLABLE), null,
		OperandTypes.or(
			OperandTypes.family(SqlTypeFamily.CHARACTER, SqlTypeFamily.CHARACTER),
			OperandTypes.family(SqlTypeFamily.CHARACTER, SqlTypeFamily.CHARACTER, SqlTypeFamily.ANY)),
		SqlFunctionCategory.SYSTEM);
```

In this way, the exception will be more reable, and we don't need the "hack" in `StringCallGen.scala`. 

```
ValidationException: SQL validation failed. From line 1, column 8 to line 1, column 48: Cannot apply 'JSON_EXISTS' to arguments of type 'JSON_EXISTS(<INTEGER>, <CHAR(5)>, <SYMBOL>)'. Supported form(s): 'JSON_EXISTS(<CHARACTER>, <CHARACTER>)'
'JSON_EXISTS(<CHARACTER>, <CHARACTER>, <ANY>)'
```

libenchao commented on flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/functions/sql/FlinkSqlOperatorTable.java:
```diff
@@ -1012,4 +1012,12 @@ public SqlMonotonicity getMonotonicity(SqlOperatorBinding call) {
+	public static final SqlFunction JSON_EXISTS = new SqlFunction(
+		"JSON_EXISTS",
+		SqlKind.OTHER_FUNCTION,
+		ReturnTypes.cascade(ReturnTypes.BOOLEAN, SqlTypeTransforms.FORCE_NULLABLE), null,
```
we should keep each param one line.

libenchao commented on flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/calls/BuiltInMethods.scala:
```diff
@@ -23,8 +23,8 @@ import org.apache.flink.table.runtime.functions._
 import org.apache.calcite.avatica.util.TimeUnitRange
 import org.apache.calcite.linq4j.tree.Types
-import org.apache.calcite.runtime.SqlFunctions
-
```
keep the original blank line.

libenchao commented on flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/calls/StringCallGen.scala:
```diff
@@ -28,11 +28,9 @@ import org.apache.flink.table.planner.functions.sql.FlinkSqlOperatorTable._
 import org.apache.flink.table.runtime.functions.SqlFunctionUtils
 import org.apache.flink.table.runtime.typeutils.TypeCheckUtils.{isCharacterString, isTimestamp, isTimestampWithLocalZone}
 import org.apache.flink.table.types.logical._
-
```
keep the blank line +1, and same for below.

libenchao commented on flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/JsonFunctionsTest.scala:
```diff
@@ -117,12 +131,48 @@ class JsonFunctionsTest extends ExpressionTestBase {
+    expectedException.expect(classOf[ValidationException])
+    expectedException.expectMessage(startsWith(expectedMessage))
+    testSqlApi(sqlExpr, "null")
+    expectedException = ExpectedException.none()
```
have you verified this test? I seems that if we encounter one exception, it will abort the test no matter it satisfies the expected exception.

XuQianJin-Stars replied:
> have you verified this test? I seems that if we encounter one exception, it will abort the test no matter it satisfies the expected exception.

Well, I tested this place. no problem.

XuQianJin-Stars replied:
> have you verified this test? I seems that if we encounter one exception, it will abort the test no matter it satisfies the expected exception.

you are right, Well,  I have changed it.

libenchao commented on flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/JsonFunctionsTest.scala:
```diff
@@ -103,9 +117,9 @@ class JsonFunctionsTest extends ExpressionTestBase {
-  ): Unit = {
+       candidate: String,
+       expectedException: Class[T]
+     ): Unit = {
```
why add this indent？

XuQianJin-Stars replied:
> why add this indent？

My idea format problem, I restore it.

libenchao commented on flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/utils/ExpressionTestBase.scala:
```diff
@@ -77,7 +77,7 @@ abstract class ExpressionTestBase {
 
   // used for accurate exception information checking.
-  val expectedException: ExpectedException = ExpectedException.none()
+  var expectedException: ExpectedException = ExpectedException.none()
```
do we still need to change this to `var`?

XuQianJin-Stars replied:
> do we still need to change this to `var`?

I have changed it.

XuQianJin-Stars commented:
> @XuQianJin-Stars Thanks for the updating. The changes LGTM now.
> CC @wuchong

hi @libenchao Thank you very much for review this pr.

twalthr commented:
Finalized in #16691.
